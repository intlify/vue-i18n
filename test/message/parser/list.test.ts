import {
  createParser,
  NodeTypes,
  MessageNode
} from '../../../src/message/parser'

let spy
beforeEach(() => {
  spy = jest.fn()
})

describe('basic', () => {
  test(`hi {0} !`, () => {
    const text = `hi {0} !`
    const parser = createParser({ onError: spy })
    const ast = parser.parse(text)

    expect(ast).toMatchSnapshot()
    expect(spy).not.toHaveBeenCalled()

    expect(ast.type).toEqual(NodeTypes.Resource)
    expect(ast.body.type).toEqual(NodeTypes.Message)
    const message = ast.body as MessageNode
    expect(message.items).toHaveLength(3)
    expect(message.items).toMatchObject([
      {
        type: NodeTypes.Text,
        value: 'hi '
      },
      {
        type: NodeTypes.List,
        index: 0
      },
      {
        type: NodeTypes.Text,
        value: ' !'
      }
    ])
  })
})

describe('multiline', () => {
  /* eslint-disable no-irregular-whitespace */
  test(`{0} {1}　{2}`, () => {
    const text = `{0} {1}　{2}`
    const parser = createParser({ onError: spy })
    const ast = parser.parse(text)

    expect(ast).toMatchSnapshot()
    expect(spy).not.toHaveBeenCalled()

    expect(ast.type).toEqual(NodeTypes.Resource)
    expect(ast.body.type).toEqual(NodeTypes.Message)
    const message = ast.body as MessageNode
    expect(message.items).toHaveLength(5)
    expect(message.items).toMatchObject([
      {
        type: NodeTypes.List,
        index: 0
      },
      {
        type: NodeTypes.Text,
        value: ' '
      },
      {
        type: NodeTypes.List,
        index: 1
      },
      {
        type: NodeTypes.Text,
        value: '　'
      },
      {
        type: NodeTypes.List,
        index: 2
      }
    ])
  })
  /* eslint-enable no-irregular-whitespace */
})

describe('whitespaces in placeholder', () => {
  test(`hi {  -1 } !`, () => {
    const text = `hi {  -1 } !`
    const parser = createParser({ onError: spy })
    const ast = parser.parse(text)

    expect(ast).toMatchSnapshot()
    expect(spy).not.toHaveBeenCalled()

    expect(ast.type).toEqual(NodeTypes.Resource)
    expect(ast.body.type).toEqual(NodeTypes.Message)
    const message = ast.body as MessageNode
    expect(message.items).toHaveLength(3)
    expect(message.items).toMatchObject([
      {
        type: NodeTypes.Text,
        value: 'hi '
      },
      {
        type: NodeTypes.List,
        index: -1
      },
      {
        type: NodeTypes.Text,
        value: ' !'
      }
    ])
  })
})

describe('new lines in placeholder', () => {
  test(`{0}\n{1}\n{2}`, () => {
    const text = `{0}\n{1}\n{2}`
    const parser = createParser({ onError: spy })
    const ast = parser.parse(text)

    expect(ast).toMatchSnapshot()
    expect(spy).not.toHaveBeenCalled()

    expect(ast.type).toEqual(NodeTypes.Resource)
    expect(ast.body.type).toEqual(NodeTypes.Message)
    const message = ast.body as MessageNode
    expect(message.items).toHaveLength(5)
    expect(message.items).toMatchObject([
      {
        type: NodeTypes.List,
        index: 0
      },
      {
        type: NodeTypes.Text,
        value: '\n'
      },
      {
        type: NodeTypes.List,
        index: 1
      },
      {
        type: NodeTypes.Text,
        value: '\n'
      },
      {
        type: NodeTypes.List,
        index: 2
      }
    ])
  })
})

test.todo(`hi {{0}} !`)
test.todo(`hi { { 0 } } !`)
test.todo(`hi {0 !`)
test.todo(`hi {  0 !`)
