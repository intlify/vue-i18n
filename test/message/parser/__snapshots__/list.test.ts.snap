// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic hi {0} ! 1`] = `
Object {
  "body": Object {
    "end": 8,
    "items": Array [
      Object {
        "end": 3,
        "loc": Object {
          "end": Object {
            "column": 4,
            "line": 1,
            "offset": 3,
          },
          "start": Object {
            "column": 1,
            "line": 1,
            "offset": 0,
          },
        },
        "start": 0,
        "type": 3,
        "value": "hi ",
      },
      Object {
        "end": 6,
        "index": 0,
        "loc": Object {
          "end": Object {
            "column": 7,
            "line": 1,
            "offset": 6,
          },
          "start": Object {
            "column": 4,
            "line": 1,
            "offset": 3,
          },
        },
        "start": 3,
        "type": 5,
      },
      Object {
        "end": 8,
        "loc": Object {
          "end": Object {
            "column": 9,
            "line": 1,
            "offset": 8,
          },
          "start": Object {
            "column": 7,
            "line": 1,
            "offset": 6,
          },
        },
        "start": 6,
        "type": 3,
        "value": " !",
      },
    ],
    "loc": Object {
      "end": Object {
        "column": 9,
        "line": 1,
        "offset": 8,
      },
      "start": Object {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
    },
    "start": 0,
    "type": 2,
  },
  "end": 8,
  "loc": Object {
    "end": Object {
      "column": 9,
      "line": 1,
      "offset": 8,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "start": 0,
  "type": 0,
}
`;

exports[`multiline {0} {1}　{2} 1`] = `
Object {
  "body": Object {
    "end": 11,
    "items": Array [
      Object {
        "end": 3,
        "index": 0,
        "loc": Object {
          "end": Object {
            "column": 4,
            "line": 1,
            "offset": 3,
          },
          "start": Object {
            "column": 1,
            "line": 1,
            "offset": 0,
          },
        },
        "start": 0,
        "type": 5,
      },
      Object {
        "end": 4,
        "loc": Object {
          "end": Object {
            "column": 5,
            "line": 1,
            "offset": 4,
          },
          "start": Object {
            "column": 4,
            "line": 1,
            "offset": 3,
          },
        },
        "start": 3,
        "type": 3,
        "value": " ",
      },
      Object {
        "end": 7,
        "index": 1,
        "loc": Object {
          "end": Object {
            "column": 8,
            "line": 1,
            "offset": 7,
          },
          "start": Object {
            "column": 5,
            "line": 1,
            "offset": 4,
          },
        },
        "start": 4,
        "type": 5,
      },
      Object {
        "end": 8,
        "loc": Object {
          "end": Object {
            "column": 9,
            "line": 1,
            "offset": 8,
          },
          "start": Object {
            "column": 8,
            "line": 1,
            "offset": 7,
          },
        },
        "start": 7,
        "type": 3,
        "value": "　",
      },
      Object {
        "end": 11,
        "index": 2,
        "loc": Object {
          "end": Object {
            "column": 12,
            "line": 1,
            "offset": 11,
          },
          "start": Object {
            "column": 9,
            "line": 1,
            "offset": 8,
          },
        },
        "start": 8,
        "type": 5,
      },
    ],
    "loc": Object {
      "end": Object {
        "column": 12,
        "line": 1,
        "offset": 11,
      },
      "start": Object {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
    },
    "start": 0,
    "type": 2,
  },
  "end": 11,
  "loc": Object {
    "end": Object {
      "column": 12,
      "line": 1,
      "offset": 11,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "start": 0,
  "type": 0,
}
`;

exports[`new lines in placeholder {0}
{1}
{2} 1`] = `
Object {
  "body": Object {
    "end": 12,
    "items": Array [
      Object {
        "end": 3,
        "index": 0,
        "loc": Object {
          "end": Object {
            "column": 4,
            "line": 1,
            "offset": 3,
          },
          "start": Object {
            "column": 1,
            "line": 1,
            "offset": 0,
          },
        },
        "start": 0,
        "type": 5,
      },
      Object {
        "end": 4,
        "loc": Object {
          "end": Object {
            "column": 1,
            "line": 2,
            "offset": 4,
          },
          "start": Object {
            "column": 4,
            "line": 1,
            "offset": 3,
          },
        },
        "start": 3,
        "type": 3,
        "value": "
",
      },
      Object {
        "end": 7,
        "index": 1,
        "loc": Object {
          "end": Object {
            "column": 4,
            "line": 2,
            "offset": 7,
          },
          "start": Object {
            "column": 1,
            "line": 2,
            "offset": 4,
          },
        },
        "start": 4,
        "type": 5,
      },
      Object {
        "end": 9,
        "loc": Object {
          "end": Object {
            "column": 1,
            "line": 3,
            "offset": 9,
          },
          "start": Object {
            "column": 4,
            "line": 2,
            "offset": 7,
          },
        },
        "start": 7,
        "type": 3,
        "value": "
",
      },
      Object {
        "end": 12,
        "index": 2,
        "loc": Object {
          "end": Object {
            "column": 4,
            "line": 3,
            "offset": 12,
          },
          "start": Object {
            "column": 1,
            "line": 3,
            "offset": 9,
          },
        },
        "start": 9,
        "type": 5,
      },
    ],
    "loc": Object {
      "end": Object {
        "column": 4,
        "line": 3,
        "offset": 12,
      },
      "start": Object {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
    },
    "start": 0,
    "type": 2,
  },
  "end": 12,
  "loc": Object {
    "end": Object {
      "column": 4,
      "line": 3,
      "offset": 12,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "start": 0,
  "type": 0,
}
`;

exports[`whitespaces in placeholder hi {  -1 } ! 1`] = `
Object {
  "body": Object {
    "end": 12,
    "items": Array [
      Object {
        "end": 3,
        "loc": Object {
          "end": Object {
            "column": 4,
            "line": 1,
            "offset": 3,
          },
          "start": Object {
            "column": 1,
            "line": 1,
            "offset": 0,
          },
        },
        "start": 0,
        "type": 3,
        "value": "hi ",
      },
      Object {
        "end": 10,
        "index": -1,
        "loc": Object {
          "end": Object {
            "column": 11,
            "line": 1,
            "offset": 10,
          },
          "start": Object {
            "column": 4,
            "line": 1,
            "offset": 3,
          },
        },
        "start": 3,
        "type": 5,
      },
      Object {
        "end": 12,
        "loc": Object {
          "end": Object {
            "column": 13,
            "line": 1,
            "offset": 12,
          },
          "start": Object {
            "column": 11,
            "line": 1,
            "offset": 10,
          },
        },
        "start": 10,
        "type": 3,
        "value": " !",
      },
    ],
    "loc": Object {
      "end": Object {
        "column": 13,
        "line": 1,
        "offset": 12,
      },
      "start": Object {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
    },
    "start": 0,
    "type": 2,
  },
  "end": 12,
  "loc": Object {
    "end": Object {
      "column": 13,
      "line": 1,
      "offset": 12,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "start": 0,
  "type": 0,
}
`;
